# Проект по SQL

<b>Цель</b>: проанализировать базу данных крупного сервиса для чтения книг по подписке с целью формирования ценностьного предложения для продукта. 

База данных содержит информацию о книгах, издательствах, авторах, а также пользовательские обзоры книг. 

Задачи:
1. Посчитайте, сколько книг вышло после 1 января 2000 года;
2. Для каждой книги посчитайте количество обзоров и среднюю оценку;
3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Знакомство с данными

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [9]:
# создаю функцию для просмотра таблиц
def result():
    table = ['books', 'authors', 'publishers', 'ratings', 'reviews']
    for t in table:
        query = f"SELECT * FROM {t}"
        print('\033[1m' + f'Таблица {t}:' + '\033[0m')
        data = pd.io.sql.read_sql(query, con = engine) 
        display(data.head(5))
        print('\033[1m' + f'Информация о {t}:' + '\033[0m')
        display(data.info())

In [10]:
result()

Таблица books:


book_id  author_id                                              title   
0        1        546                                       'Salem's Lot  \
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Информация о books:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

Таблица authors:


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Информация о authors:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

Таблица publishers:


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Информация о publishers:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

Таблица ratings:


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Информация о ratings:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

Таблица reviews:


review_id  book_id       username   
0          1        1   brandtandrea  \
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Информация о reviews:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

- Таблица books содержит id книги и автора, название, количество страниц, дату публикации и id публикации, всего 1000 строк.
- Таблица authors содержит имена и id авторов.
- Таблица publishers содержит информацию о названии издательтсв.
- Таблица ratings содержит информацию о имени пользователя, оставившем оценку, и собственно оценку.
- Таблица reviews содержит данные о пользовательских обзорах, имени автора обзора и текст обзора. 

## Книги, вышедшие после 01.01.2000

In [4]:
# создаю функцию для подключения к БД
def show_df(query):
    data_raw = pd.io.sql.read_sql(query, con = engine) 
    display(data_raw)

In [5]:
book_cnt = '''
            SELECT COUNT(*)
            FROM books
            WHERE publication_date::date > '2000-01-01'
            
            '''

show_df(book_cnt)

count
0    819

In [6]:
book_min_max = '''
                SELECT 
                    MIN(publication_date::date), 
                    MAX(publication_date::date)
                FROM books
                
                '''

show_df(book_min_max)

min         max
0  1952-12-01  2020-03-31

Из 1000 книг, представленных в датафрейме, выходивших в период с 01.12.1952 по 31.03.2020 гг, 819 книг вышло после 2000 года. То есть большинство книг в нашем фрейме современные, вышедшие уже в 21 веке. 

## Количество обзоров и средняя оценка

In [7]:
books_reviews = ''' WITH books_rewies AS 
                            (SELECT  
                                b.book_id, 
                                b.title, 
                                COUNT(review_id) AS reviews_count
                            FROM books AS b
                            LEFT JOIN reviews AS re ON re.book_id = b.book_id
                            GROUP BY b.book_id),
                            
                        books_ratings AS
                            (SELECT  
                                b.book_id, 
                                ROUND(AVG(ra.rating), 2) AS rating_avg
                            FROM ratings AS ra 
                            LEFT JOIN books AS b ON ra.book_id = b.book_id
                            GROUP BY b.book_id)
                            
                    SELECT * 
                    FROM books_rewies JOIN books_ratings
                    ON books_rewies.book_id = books_ratings.book_id
                    ORDER BY reviews_count DESC
                '''

show_df(books_reviews)

book_id                                              title   
0        948                            Twilight (Twilight  #1)  \
1        963                                Water for Elephants   
2        734                                   The Glass Castle   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...   
4        695  The Curious Incident of the Dog in the Night-Time   
..       ...                                                ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel   
996      808                            The Natural Way to Draw   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites   
998      221                          Essential Tales and Poems   
999      191  Disney's Beauty and the Beast (A Little Golden...   

     reviews_count  rating_avg  
0                7        3.66  
1                6        3.98  
2                6        4.21  
3                6        4.41  
4                6        4.08  
..             ...         ...  
995              0        3.67  
996              0        3.00  
997              0        5.00  
998              0        4.00  
999              0        4.00  

[1000 rows x 4 columns]

В таблице можно увидеть количество отзывов на каждую книгу и ее средний рейтинг. Наибольшее количество отзывов 7 получила книга Сумерки с средней оценкой 3,66.

In [12]:
books_reviews = ''' WITH books_rewies AS 
                            (SELECT  
                                b.book_id, 
                                b.title, 
                                COUNT(review_id) AS reviews_count
                            FROM books AS b
                            LEFT JOIN reviews AS re ON re.book_id = b.book_id
                            GROUP BY b.book_id),
                            
                        books_ratings AS
                            (SELECT  
                                b.book_id, 
                                b.title, 
                                ROUND(AVG(ra.rating), 2) AS rating_avg
                            FROM ratings AS ra 
                            LEFT JOIN books AS b ON ra.book_id = b.book_id
                            GROUP BY b.book_id)
                            
                    SELECT 
                        AVG(reviews_count) AS reviews_avg, 
                        AVG(rating_avg) AS  rating_avg
                    FROM books_rewies JOIN books_ratings
                    ON books_rewies.book_id = books_ratings.book_id
                '''

show_df(books_reviews)

reviews_avg  rating_avg
0        2.793     3.89904

В среднем на книгу приходится 2 отзыва. Средняя оценка за книгу 3.9.

## Издательство, которое выпустило наибольшее количество книг

In [13]:
publishers = ''' 
                SELECT 
                    publisher, 
                    COUNT(book_id) AS books_count
                FROM books AS b LEFT JOIN publishers AS p ON b.publisher_id = p.publisher_id
                WHERE num_pages > 50
                GROUP BY publisher
                ORDER BY 2 DESC
                LIMIT 1
                
              '''

show_df(publishers)

publisher  books_count
0  Penguin Books           42

Наибольшее количество книг с объемом страниц более 50 выпустило издательство под названием Penguin Books.

## Автор с самой высокой средней оценкой 

In [14]:
authors =       '''       
                    WITH rating_top AS
                        (SELECT 
                               b.book_id, 
                               b.author_id,
                               b.title, 
                               COUNT(b.book_id) AS rating_count, 
                               AVG(ra.rating) AS rating_avg
                               
                     FROM books AS b JOIN ratings AS ra ON ra.book_id = b.book_id
                     GROUP BY b.book_id
                     HAVING COUNT(b.book_id) > 50
                     ORDER BY 4 DESC),
                        
                    rating_avg AS
                        (SELECT 
                        author, 
                        title, 
                        ROUND(rating_avg, 2) AS rating_avg
                    FROM rating_top JOIN authors AS au ON au.author_id = rating_top.author_id
                    ORDER BY 3 DESC)
                    
                    SELECT
                        rating_avg.author,
                        AVG(rating_avg)
                    FROM  rating_avg
                    GROUP BY author
                    LIMIT 1
                      

                '''

show_df(authors)

author    avg
0  J.K. Rowling/Mary GrandPré  4.285

Автор с более высокой средней оценкой из популярных книг (тех, у которых количество оценок превышает 50 за книгу) - Джоан Роулинг.

## Среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [15]:
users =       '''      
                        WITH 
                        users AS
                            (SELECT username
                            FROM  ratings AS ra 
                            GROUP BY username
                            HAVING COUNT(rating_id) > 50), 
                            
                        reviews AS
                            (SELECT re.username, COUNT(re.review_id) AS reviews_count
                            FROM reviews AS re
                            JOIN users ON re.username = users.username
                            GROUP BY re.username)
                        
                        SELECT ROUND(AVG(reviews_count)) AS reviews_avg
                        FROM reviews
                        
                '''
show_df(users)

reviews_avg
0         24.0

В среднем активные пользователи (которые ставят более 50 оценок) пишут 24 обзора на книги. 

# Вывод

Были проанализированы базы данных крупного сервиса по чтению книг по подписке. 

В БД находится информация о 1000 различных книг, выпущенных в период с 01.12.1952 по 31.03.2020 гг, из которых 819 книг вышло после 2000 года, то есть большинство книг были написаны уже в 21 веке. 
- Для каждой книги был подсчитан средний рейтинг и количество обзоров на нее, наибольшее количество обзоров определяется у книги Сумерки с рейтингом 3.6. В среднем на книгу приходится 2 отзыва. Средняя оценка за книгу 3.9.
- Самое популярное издательство, которое напечатало 42 книги за весь период, Penguin Books.
- Автор с наболее высокой оценкой среди популярных книг (тех, у которых количество оценок превышает 50 за книгу) - Джоан Роулинг.
- В среднем активные пользователи, которые ставят более 50 оценок, пишут 24 обзора на книги.